File: D1L1/Spamatronic.cpp
#include <cctype>
#include <vector>
#include <string>
#include <set>
#include <sstream>
using namespace std;

void tokenize(set<string> &tokens, string const &st) {
    ostringstream oss;
    for (int i = 0; i != st.size(); ++i)
        if (isalpha(st[i]))
            oss << toupper(st[i]);
        else if (!oss.str().empty()) {
            tokens.insert(oss.str());
            oss.str("");
        }
    if (!oss.str().empty()) 
        tokens.insert(oss.str());
} 
vector<int> filter(vector<string> const &knownSpam,
                   vector<string> const &newMail) {
    vector<int> ret; set<string> spams;
    for (auto it = knownSpam.begin(); it != knownSpam.end(); ++it)
        tokenize(spams, *it);
    for (int i = 0; i != newMail.size(); ++i) {
        set<string> tokens; tokenize(tokens, newMail[i]); int k = 0;
        for (auto it = tokens.begin(); it != tokens.end(); ++it)
            if (spams.find(*it) != spams.end()) 
                ++k;
        if (tokens.size() * 3 <= k * 4) 
            for (auto it = tokens.begin(); it != tokens.end(); ++it) 
                spams.insert(*it);
        else ret.push_back(i);
    }
    return ret;
}

File: D1L1/Spamatronic.hs
module Spamatronic where
import Data.Char
import Data.Set (Set)
import qualified Data.Set as Set
import Data.List.Split

tokenize :: String -> Set String
tokenize = foldl (flip Set.insert) Set.empty . 
           map (map toUpper) . wordsBy (not . isLetter)

isSpam :: Set String -> Set String -> Bool
isSpam inner outer = 3 * n <= 4 * k
    where k = Set.size $ Set.intersection inner outer
          n = Set.size inner

filterSpam :: [String] -> [String] -> [Int]
filterSpam knownSpam newMail = 
    snd . foldl pair (Set.unions . map tokenize $ knownSpam, []) . 
    zip (iterate (+1) 0) . map tokenize $ newMail
    where pair (spams, ret) (i, mail) = 
              if isSpam mail spams then (Set.union mail spams, ret)
                                   else (spams, ret ++ [i])